vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_flash.c

```diff
@@ -0,0 +1,490 @@
+{
+    return ( pxFlashHandle != NULL ) && ( pxFlashHandle->ucState == IOT_FLASH_OPENED );
+}
+
+/*---------------------------------------------------------------------------------*
+*                               API Implementation                                *
+*---------------------------------------------------------------------------------*/
+IotFlashHandle_t iot_flash_open( int32_t lFlashInstance )
+{
+    IotFlashHandle_t xHandle = NULL;
+
+    if( ( lFlashInstance == 0 ) && ( xFlashDesc.ucState == IOT_FLASH_CLOSED ) )
+    {
+        /* Calculate flash addresses */
+        prvGetFlashInfo();
+        #ifdef SOFTDEVICE_PRESENT
+
+            /* Initialize an fstorage instance using the nrf_fstorage_sd backend.
+             * nrf_fstorage_sd uses the SoftDevice to write to flash. This implementation can safely be
+             * used whenever there is a SoftDevice, regardless of its status (enabled/disabled). */
+            nrf_fstorage_api_t * p_fs_api = &nrf_fstorage_sd;
+            IotLogDebug( "Backend: nrf_fstorage_sd" );
+        #else
+
+            /* Initialize an fstorage instance using the nrf_fstorage_nvmc backend.
+             * nrf_fstorage_nvmc uses the NVMC peripheral. This implementation can be used when the
+             * SoftDevice is disabled or not present.
+             *
+             * Using this implementation when the SoftDevice is enabled results in a hardfault. */
+            nrf_fstorage_api_t * p_fs_api = &nrf_fstorage_nvmc;
+            IotLogDebug( "Backend: nrf_fstorage_nvmc" );
+        #endif
+
+        xFlashDesc = DEFAULT_FLASH_DESC;
+        ret_code_t xReturnCode_NRF = nrf_fstorage_init( &iot_fstorage, p_fs_api, NULL );
+
+        if( xReturnCode_NRF == NRF_SUCCESS )
+        {
+            xFlashDesc.xFlashInfo.ulFlashSize = iot_fstorage.end_addr - iot_fstorage.start_addr + 1;
+            xFlashDesc.ucState = IOT_FLASH_OPENED;
+            xHandle = &xFlashDesc;
+        }
+        else
+        {
+            IotLogError( "%s: Internal NRF error. Could not open handle[%d]", __func__, lFlashInstance );
+        }
+    }
+
+    return xHandle;
+}
+
+IotFlashInfo_t * iot_flash_getinfo( IotFlashHandle_t const pxFlashHandle )
+{
+    IotFlashInfo_t * pxFlashInfo = NULL;
+
+    if( prvIsOperableHandle( pxFlashHandle ) )
+    {
+        pxFlashInfo = &xFlashDesc.xFlashInfo;
+    }
+
+    return pxFlashInfo;
+}
+
+void iot_flash_set_callback( IotFlashHandle_t const pxFlashHandle,
+                             IotFlashCallback_t xCallback,
+                             void * pvUserContext )
+{
+    if( prvIsOperableHandle( pxFlashHandle ) )
+    {
+        pxFlashHandle->xCallback = xCallback;
+        pxFlashHandle->pvUserContext = pvUserContext;
+    }
+    else
+    {
+        IotLogWarn( "%s: Null or closed IotFlashHandle_t(0x%x)", __func__, pxFlashHandle );
+    }
+}
+
+int32_t iot_flash_ioctl( IotFlashHandle_t const pxFlashHandle,
```

Comment:
I truly hate the ioctl pattern.  The cyclomatic complexity goes up and the clarity goes down.
Why can't we simply have the functions we mean to have?  Is there a limit to our function count?

Reply:
Honoring what the existing API (`iot_flash.h`) expects. The `common_io/*.h` APIs were the result of extensive discussion. @ejacobus can provide details on those discussions and resulting decisions. 

From my observation, the `ioctls` are useful for drawing a line between "standard" and "non-standard" common_io functions.

Since the `*.h` APIs predate the `*.c` implementations, any non-additive API changes could also mess with customer implementations of `iot_flash.c`, if they wrote one.